In [1]:
import matplotlib.pyplot as plt
import numpy as np

from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.layers import Conv2D, Input, Flatten, Dense
from tensorflow.keras.losses import categorical_crossentropy
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.layers import GlobalMaxPooling2D, MaxPooling2D, AveragePooling2D, GlobalAveragePooling2D 

In [2]:
(train_images,train_labels),(test_images,test_labels)=fashion_mnist.load_data()

# Normalize the images.
train_images = (train_images / 255)*2-1
test_images = (test_images / 255)*2-1

# Flatten the images.
train_images = train_images#.reshape((-1, 28*28))
test_images = test_images#.reshape((-1, 28*28))

print(train_images.shape)
print(test_images.shape) 

4423680/4422102 [==============================] - 0s 0us/step
(60000, 28, 28)
(10000, 28, 28)


In [3]:
# трансформация лейблов в one-hot encoding
train_labels = to_categorical(train_labels, 10) 
test_labels = to_categorical(test_labels, 10) 

# изменение размерности массива в 4D массив
train_images = train_images.reshape(train_images.shape[0], 28,28,1)
test_images = test_images.reshape(test_images.shape[0], 28,28,1)

print(train_images.shape) 
print(test_images.shape)  
print()
print(train_labels.shape) 
print(test_labels.shape)  

(60000, 28, 28, 1)
(10000, 28, 28, 1)

(60000, 10)
(10000, 10)


# Анализ влияния ширины сети


In [52]:
def models_width(filters_1, filters_2, epochs=5):
    model = Sequential()
    
    # первый сверточный слой
    model.add(Conv2D(filters_1, kernel_size=(5, 5), strides=(1, 1), activation='sigmoid', input_shape=(28,28,1), padding="same"))
    
    # второй пуллинговый слой
    model.add(MaxPooling2D(pool_size=(3, 3), strides=(1, 1), padding='valid'))
    
    # пятый сверточный слой
    model.add(Conv2D(filters_2, kernel_size=(5, 5), strides=(1, 1), activation='sigmoid', padding='valid'))
    
    # сглаживание CNN выхода чтобы можно было его присоединить к полносвязногому слою
    model.add(Flatten())
    
    # шестой полносвязный слой
    model.add(Dense(80, activation='sigmoid'))
    
    # выходной слой с функцией активации softmax
    model.add(Dense(10, activation='softmax'))    
    
    # компилияция модели
    model.compile(loss=categorical_crossentropy, optimizer='adam', metrics=["accuracy"])    
    
    hist = model.fit(x=train_images,
                     y=train_labels, 
                     epochs=epochs, 
                     batch_size=128, 
                     validation_data=(test_images, test_labels), 
                     verbose=1)
    
    test_score = model.evaluate(test_images, test_labels)
    
    print("Test loss {:.4f}, accuracy {:.2f}%".format(test_score[0], test_score[1] * 100))

In [67]:
models_width(2, 4, epochs=20)

Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 1.1169 - accuracy: 0.6085 - val_loss: 0.6752 - val_accuracy: 0.7513
Epoch 2/20
469/469 [==============================] - 2s 3ms/step - loss: 0.6001 - accuracy: 0.7759 - val_loss: 0.5763 - val_accuracy: 0.7808
Epoch 3/20
469/469 [==============================] - 2s 3ms/step - loss: 0.5124 - accuracy: 0.8093 - val_loss: 0.5051 - val_accuracy: 0.8118
Epoch 4/20
469/469 [==============================] - 2s 3ms/step - loss: 0.4532 - accuracy: 0.8341 - val_loss: 0.4622 - val_accuracy: 0.8281
Epoch 5/20
469/469 [==============================] - 2s 3ms/step - loss: 0.4144 - accuracy: 0.8487 - val_loss: 0.4187 - val_accuracy: 0.8493
Epoch 6/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3875 - accuracy: 0.8598 - val_loss: 0.4099 - val_accuracy: 0.8531
Epoch 7/20
469/469 [==============================] - 2s 3ms/step - loss: 0.3693 - accuracy: 0.8653 - val_loss: 0.3837 - val_accuracy: 0.8614
Epoch 

In [66]:
models_width(5, 10, epochs=20)

Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 1.0980 - accuracy: 0.6373 - val_loss: 0.6055 - val_accuracy: 0.7782
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 0.5112 - accuracy: 0.8164 - val_loss: 0.4813 - val_accuracy: 0.8233
Epoch 3/20
469/469 [==============================] - 2s 4ms/step - loss: 0.4198 - accuracy: 0.8490 - val_loss: 0.4116 - val_accuracy: 0.8513
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3762 - accuracy: 0.8648 - val_loss: 0.3843 - val_accuracy: 0.8591
Epoch 5/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3486 - accuracy: 0.8738 - val_loss: 0.3630 - val_accuracy: 0.8702
Epoch 6/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3274 - accuracy: 0.8814 - val_loss: 0.3477 - val_accuracy: 0.8733
Epoch 7/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3117 - accuracy: 0.8863 - val_loss: 0.3272 - val_accuracy: 0.8803
Epoch 

In [64]:
models_width(8, 16, epochs=20)

Epoch 1/20
469/469 [==============================] - 2s 5ms/step - loss: 1.2058 - accuracy: 0.6200 - val_loss: 0.6354 - val_accuracy: 0.7824
Epoch 2/20
469/469 [==============================] - 2s 5ms/step - loss: 0.5198 - accuracy: 0.8188 - val_loss: 0.4816 - val_accuracy: 0.8312
Epoch 3/20
469/469 [==============================] - 2s 5ms/step - loss: 0.4203 - accuracy: 0.8524 - val_loss: 0.4088 - val_accuracy: 0.8556
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3737 - accuracy: 0.8675 - val_loss: 0.3831 - val_accuracy: 0.8605
Epoch 5/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3444 - accuracy: 0.8767 - val_loss: 0.3582 - val_accuracy: 0.8714
Epoch 6/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3229 - accuracy: 0.8838 - val_loss: 0.3371 - val_accuracy: 0.8777
Epoch 7/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3026 - accuracy: 0.8903 - val_loss: 0.3304 - val_accuracy: 0.8799
Epoch 

In [71]:
models_width(10, 20, epochs=20)

Epoch 1/20
469/469 [==============================] - 2s 5ms/step - loss: 1.2737 - accuracy: 0.5742 - val_loss: 0.6350 - val_accuracy: 0.7820
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 0.5238 - accuracy: 0.8167 - val_loss: 0.4725 - val_accuracy: 0.8357
Epoch 3/20
469/469 [==============================] - 2s 4ms/step - loss: 0.4215 - accuracy: 0.8518 - val_loss: 0.4110 - val_accuracy: 0.8539
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3770 - accuracy: 0.8654 - val_loss: 0.3950 - val_accuracy: 0.8558
Epoch 5/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3454 - accuracy: 0.8760 - val_loss: 0.3653 - val_accuracy: 0.8698
Epoch 6/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3208 - accuracy: 0.8842 - val_loss: 0.3473 - val_accuracy: 0.8721
Epoch 7/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3050 - accuracy: 0.8893 - val_loss: 0.3314 - val_accuracy: 0.8815
Epoch 

In [68]:
models_width(15, 30, epochs=20)

Epoch 1/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3359 - accuracy: 0.0990 - val_loss: 2.3058 - val_accuracy: 0.1000
Epoch 2/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3046 - accuracy: 0.1014 - val_loss: 2.3046 - val_accuracy: 0.1000
Epoch 3/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3046 - accuracy: 0.0975 - val_loss: 2.3052 - val_accuracy: 0.1000
Epoch 4/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3048 - accuracy: 0.0984 - val_loss: 2.3045 - val_accuracy: 0.1000
Epoch 5/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3048 - accuracy: 0.0992 - val_loss: 2.3037 - val_accuracy: 0.1000
Epoch 6/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3044 - accuracy: 0.0986 - val_loss: 2.3044 - val_accuracy: 0.1000
Epoch 7/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3044 - accuracy: 0.0996 - val_loss: 2.3050 - val_accuracy: 0.1000
Epoch 

### РЕЗУЛЬТАТЫ:

Модель: сверточный слой - пулинг - сверотчный слой - полносвязный - выходной.

20 эпох достаточно, чтобы на всех моделях точность на валидации выходила на aсимптоту.

Рассматривалось влияние количества ядер на сверточных слоях.

* Ядер 2 и 4:   accuracy 88.80%
* Ядер 5 и 10:  accuracy 90.20%
* Ядер 8 и 16:  accuracy 90.77%

* На следующих настройках обучение ссети нестабильно, сеть начала обучаться 
после нескольких попыток перезапуска:
Ядер 10 и 20: accuracy 90.53%

* На следующих настройках сеть перестает обучаться:
Ядер 15 и 30: accuracy 10.00%

### ВЫВОД:
увеличение количества ядер может приводить к росту точности, но в какой-то 
момент начинается нестабильность обучения, увеличивается вероятность того, что сеть не сможет обучиться и в конечном счете совсем перестает обучаться

# Анализ глубины сети

In [77]:
def model_layers(n):
    # инициализация пустой модели
    model = Sequential()
    
    for _ in range(n):
        model.add(Conv2D(6, kernel_size=(5, 5), strides=(1, 1), activation='sigmoid', input_shape=(28,28,1), padding="same"))    
        model.add(AveragePooling2D(pool_size=(3, 3), strides=(1, 1), padding='valid'))
        model.add(Conv2D(6, kernel_size=(5, 5), strides=(1, 1), activation='sigmoid', padding='valid'))
    
    # сглаживание CNN выхода чтобы можно было его присоединить к полносвязногому слою
    model.add(Flatten())    
    # шестой полносвязный слой
    model.add(Dense(84, activation='sigmoid'))    
    # выходной слой с функцией активации softmax
    model.add(Dense(10, activation='softmax'))    
    # компилияция модели
    model.compile(loss=categorical_crossentropy, optimizer='adam', metrics=["accuracy"])      
    hist = model.fit(x=train_images,
                     y=train_labels, 
                     epochs=20, 
                     batch_size=128, 
                     validation_data=(test_images, test_labels), 
                     verbose=1)    
    test_score = model.evaluate(test_images, test_labels)    
    print("Test loss {:.4f}, accuracy {:.2f}%".format(test_score[0], test_score[1] * 100))

In [78]:
model_layers(1)

Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 1.0230 - accuracy: 0.6451 - val_loss: 0.6329 - val_accuracy: 0.7650
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 0.5202 - accuracy: 0.8098 - val_loss: 0.5009 - val_accuracy: 0.8219
Epoch 3/20
469/469 [==============================] - 2s 4ms/step - loss: 0.4337 - accuracy: 0.8447 - val_loss: 0.4312 - val_accuracy: 0.8453
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3930 - accuracy: 0.8596 - val_loss: 0.4114 - val_accuracy: 0.8502
Epoch 5/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3659 - accuracy: 0.8676 - val_loss: 0.3863 - val_accuracy: 0.8615
Epoch 6/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3457 - accuracy: 0.8752 - val_loss: 0.3751 - val_accuracy: 0.8635
Epoch 7/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3275 - accuracy: 0.8822 - val_loss: 0.3594 - val_accuracy: 0.8708
Epoch 

In [79]:
model_layers(2)

Epoch 1/20
469/469 [==============================] - 2s 5ms/step - loss: 1.2204 - accuracy: 0.5622 - val_loss: 0.7425 - val_accuracy: 0.7152
Epoch 2/20
469/469 [==============================] - 2s 5ms/step - loss: 0.6667 - accuracy: 0.7456 - val_loss: 0.6367 - val_accuracy: 0.7613
Epoch 3/20
469/469 [==============================] - 2s 5ms/step - loss: 0.5922 - accuracy: 0.7735 - val_loss: 0.5837 - val_accuracy: 0.7781
Epoch 4/20
469/469 [==============================] - 2s 5ms/step - loss: 0.5486 - accuracy: 0.7934 - val_loss: 0.5554 - val_accuracy: 0.7937
Epoch 5/20
469/469 [==============================] - 2s 5ms/step - loss: 0.5194 - accuracy: 0.8037 - val_loss: 0.5489 - val_accuracy: 0.7892
Epoch 6/20
469/469 [==============================] - 2s 5ms/step - loss: 0.4950 - accuracy: 0.8135 - val_loss: 0.5084 - val_accuracy: 0.8050
Epoch 7/20
469/469 [==============================] - 2s 5ms/step - loss: 0.4730 - accuracy: 0.8227 - val_loss: 0.4946 - val_accuracy: 0.8091
Epoch 

In [80]:
model_layers(3)

Epoch 1/20
469/469 [==============================] - 3s 6ms/step - loss: 1.5391 - accuracy: 0.4221 - val_loss: 0.9377 - val_accuracy: 0.6428
Epoch 2/20
469/469 [==============================] - 2s 5ms/step - loss: 0.8191 - accuracy: 0.6913 - val_loss: 0.7699 - val_accuracy: 0.7133
Epoch 3/20
469/469 [==============================] - 2s 5ms/step - loss: 0.7103 - accuracy: 0.7292 - val_loss: 0.6976 - val_accuracy: 0.7397
Epoch 4/20
469/469 [==============================] - 2s 5ms/step - loss: 0.6578 - accuracy: 0.7484 - val_loss: 0.6655 - val_accuracy: 0.7559
Epoch 5/20
469/469 [==============================] - 2s 5ms/step - loss: 0.6246 - accuracy: 0.7635 - val_loss: 0.6311 - val_accuracy: 0.7658
Epoch 6/20
469/469 [==============================] - 2s 5ms/step - loss: 0.5950 - accuracy: 0.7763 - val_loss: 0.6164 - val_accuracy: 0.7737
Epoch 7/20
469/469 [==============================] - 2s 5ms/step - loss: 0.5702 - accuracy: 0.7871 - val_loss: 0.5829 - val_accuracy: 0.7877
Epoch 

In [81]:
model_layers(4)

Epoch 1/20
469/469 [==============================] - 3s 6ms/step - loss: 2.2560 - accuracy: 0.1224 - val_loss: 1.6741 - val_accuracy: 0.3359
Epoch 2/20
469/469 [==============================] - 3s 6ms/step - loss: 1.4370 - accuracy: 0.4134 - val_loss: 1.3463 - val_accuracy: 0.4743
Epoch 3/20
469/469 [==============================] - 3s 6ms/step - loss: 1.2097 - accuracy: 0.5327 - val_loss: 1.1881 - val_accuracy: 0.5464
Epoch 4/20
469/469 [==============================] - 3s 6ms/step - loss: 1.1064 - accuracy: 0.5763 - val_loss: 1.0721 - val_accuracy: 0.6016
Epoch 5/20
469/469 [==============================] - 3s 6ms/step - loss: 1.0022 - accuracy: 0.6234 - val_loss: 0.9821 - val_accuracy: 0.6312
Epoch 6/20
469/469 [==============================] - 3s 6ms/step - loss: 0.9494 - accuracy: 0.6418 - val_loss: 0.9416 - val_accuracy: 0.6465
Epoch 7/20
469/469 [==============================] - 3s 6ms/step - loss: 0.9086 - accuracy: 0.6540 - val_loss: 0.9018 - val_accuracy: 0.6569
Epoch 

### Результаты:

Первая модель: сверточный слой - пулинг - сверотчный слой - полносвязный - выходной.

Для каждой новой модели добавляются дополнительно по 2 слоя - пулинг и сверточный перед полносвязыным. Все сверотчные и пулинг слои одинаковы по настройкам.

* модель 1 (2 сверотчных и 1 пулинг) accuracy 89.03%
* модель 2 (3 сверотчных и 2 пулинг) accuracy 86.20%
* модель 3 (4 сверотчных и 3 пулинг) accuracy 84.24%
* модель 4 (5 сверотчных и 4 пулинг) accuracy 76.16% 

### Вывод 
простое добавление идентичных слоев уменьшает точность модели

# Общий вывод
Для оптимальной модели необоходимо строить архитектуру сети: подбирать оптимальное количство слоев и оптимальное количество ядер в каждом слое